### DISCUSSION TITLE
EventCatcher widget - use setPointerCapture?

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#5937] [Ideas]
We could use `setPointerCapture` (https://developer.mozilla.org/en-US/docs/Web/API/Element/setPointerCapture) as an additional parameter for the eventcatcher widget.

I would like to discuss this and the benefits it would give us

**Comment by Jermolene** [comment]
Hi @BurningTreeC good idea. The challenge is that the eventcatcher widget might get destroyed and recreated when a parent widget is refreshed. Presumably we'd have to track the current capturing element rather like the work you've done on tracking the focussed element.

**Reply by BurningTreeC** [reply]
>Presumably we'd have to track the current capturing element rather like the work you've done on tracking the focussed element.

sounds complicated @Jermolene

**Reply by BurningTreeC** [reply]
isn't it enough to `domNode.setPointerCapture(event.pointerId)` ?

**Reply by Jermolene** [reply]
The trouble is that a refresh might happen while the pointer is captured, and during the refresh the pointer capture would (presumably) get released prematurely when the DOM node is destroyed.

**Reply by BurningTreeC** [reply]
I am able to work around that @Jermolene

**Reply by BurningTreeC** [reply]
... if we make it possible to dynamically assign styles to the domNode created by the eventcatcher widget

**Reply by Jermolene** [reply]
Great, can you outline the approach you're thinking of?

**Reply by BurningTreeC** [reply]
What I have in mind is a `splitter-panel` , a macro that creates a variable number of flexbox columns which can be resized by dragging. Therefor I create `divs` using the `eventcatcher` widget whilst every `eventcatcher` widget listens on its own created domNode for `pointer events`. I'm able to create all those necessary `divs` down to the flexbox "items" which are being resized without refreshing widgets during the drag. The styles for the flexbox "items" are calculated and applied inline.

Now this can work without editing the `$:/core/ui/PageTemplate` tiddler if we use `setPointerCapture(event.pointerId)` on the `eventcatcher domNode` and if we can apply styles to the `eventcatcher domNode` dynamically

**Reply by Jermolene** [reply]
Hi @BurningTreeC indeed, the simple approach will work in situations where no refreshing occurs while the pointer is captured.

The `<$edit-bitmap>` widget suffers from the same problem: it would misbehave if it is regenerated while drawing a stroke.

**Reply by BurningTreeC** [reply]
I have built this prototype of a `splitterpanel` as a `WikiText macro` ... it still has some quirks and some handling is missing, but I believe that if we add `setPointerCapture` and the dynamic handling of styles to the `<$eventcatcher>` widget we can make such a `splitterpanel` reality

https://user-images.githubusercontent.com/9407182/128591832-5ff1cebe-b1e2-4dd5-ae4d-b86ec1bc9aa8.mp4

**Reply by BurningTreeC** [reply]
@Jermolene, @pmario, @saqimtiaz ... any thoughts about such a splitterpanel macro?

**Reply by Jermolene** [reply]
Hi @BurningTreeC very interesting! Does this still use the dynamic stylesheet technique we saw before?

**Reply by BurningTreeC** [reply]
No @Jermolene , this uses inline styles

**Reply by Jermolene** [reply]
Great, let's see it!

**Reply by BurningTreeC** [reply]
I'm a bit shy about showing this macro :grin: ... it only works if we don't change the width of the container and if we don't change the amount of columns. If we change those, we need to delete the state tiddler.

I'm still working on this macro and the handling for when the width and the amount of columns changes is missing.

But I guess some more pairs of eyes on the code would help

**Reply by Jermolene** [reply]
> I'm a bit shy about showing this macro 😁 ... it only works if we don't change the width of the container and if we don't change the amount of columns. If we change those, we need to delete the state tiddler.
> 
> I'm still working on this macro and the handling for when the width and the amount of columns changes is missing.
> 
> But I guess some more pairs of eyes on the code would help

Inevitably, I'm always very focussed on the question of "how do we get this into the core"; an alternative approach would be to work on getting it packaged up for users as it is so that you can explore with them how much these limitations matter in practice...
